### PR TITLE
IGNITE-13001 .NET: Add Thin Client Compute

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Add `IComputeClient` with `ExecuteJavaTask`
* Add `WithKeepBinary` flag to compute protocol
* Improve `unknown pair` exceptions for thin and thick clients
* Add new benchmark project based on BenchmarkDotNet library
* Fix cluster operation feature flags handling

### CODE REVIEW

File: modules/platforms/dotnet/Apache.Ignite.Core/Client/Compute/IComputeClient.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    public interface IComputeClient
+    {
+        /// <summary>
+        /// Executes Java task by class name.
+        /// </summary>
+        /// <param name="taskName">Java task name.</param>
+        /// <param name="taskArg">Optional argument of task execution, can be null.</param>
+        /// <returns>Task result.</returns>
+        /// <typeparam name="TRes">Type of task result.</typeparam>
+        TRes ExecuteJavaTask<TRes>(string taskName, object taskArg);
```
Comment (isapego) [inline modules/platforms/dotnet/Apache.Ignite.Core/Client/Compute/IComputeClient.cs:37]: This is the only method of the class which is tested. Add at least tests for Async execution, WithKeepBinary...

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Compute/ComputeClientDisabledTests.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    /// <summary>
+    /// Tests client compute with disabled server-side settings.
+    /// </summary>
+    public class ComputeClientDisabledTests : ClientTestBase
+    {
+        /// <summary>
+        /// Tests that Compute throws correct exception when not enabled on server.
+        /// </summary>
+        [Test]
+        public void TestComputeThrowsCorrectExceptionWhenNotEnabledOnServer()
```
Comment (isapego) [inline modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Compute/ComputeClientDisabledTests.cs:34]: Can you add a more description in Doxy comment what do this test does?

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientSocket.cs [hunk 3]
```diff
@@ -66,7 +66,7 @@ internal sealed class ClientSocket : IDisposable
 
         /** Current version. */
         public static readonly ClientProtocolVersion CurrentProtocolVersion = Ver170;
-
+
```
Comment (isapego) [inline modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientSocket.cs:69]: Whitespace :)

File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Compute/ComputeClientDisabledTests.cs (hunk 2 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Compute/ComputeClientDisabledTests.cs:34]: It is already there: `Tests that Compute throws correct exception when not enabled on server.`. What would you add to this description?

File: modules/platforms/dotnet/Apache.Ignite.Core/Client/Compute/IComputeClient.cs (hunk 1 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Client/Compute/IComputeClient.cs:37]: Please see `ComputeClientTests` class, it covers all APIs and different use cases, exceptions, flags, timeouts.

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientSocket.cs (hunk 3 above)
Comment (ptupitsyn) [reply modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientSocket.cs:69]: Fixed, thanks :)

File: modules/core/src/test/java/org/apache/ignite/internal/client/thin/TestTask.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ */
+@ComputeTaskName(ComputeTaskTest.TEST_TASK_NAME)
+public class TestTask extends ComputeTaskAdapter<Long, T2<UUID, Set<UUID>>> {
+    /** Ignite. */
+    @IgniteInstanceResource
+    Ignite ignite;
+
+    /** {@inheritDoc} */
+    @Override public @NotNull Map<? extends ComputeJob, ClusterNode> map(List<ClusterNode> subgrid,
+                                                                         @Nullable Long arg) throws IgniteException {
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/core/src/test/java/org/apache/ignite/internal/client/thin/TestTask.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/compute/ClientExecuteTaskRequest.java [hunk 5]
```diff
@@ -77,7 +79,15 @@ public ClientExecuteTaskRequest(BinaryRawReader reader) {
         long taskId = ctx.resources().put(task);
 
         try {
-            task.execute(taskId, taskName, arg, nodeIds, flags, timeout);
+            Object arg0 = arg;
+
+            // Deserialize as part of process() call - not in constructor - for proper error handling.
+            // Failure to deserialize binary object should not be treated as a failure to decode request.
+            if ((flags & ClientComputeTask.KEEP_BINARY_FLAG_MASK) == 0 && arg instanceof BinaryObject) {
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/compute/ClientExecuteTaskRequest.java:86]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/MarshallerContextImpl.java [hunk 6]
```diff
@@ -391,34 +411,56 @@ public void onMappingAccepted(final MarshallerMappingItem item) {
...
-                    if (mappedName == null) {
-                        GridFutureAdapter<MappingExchangeResult> fut = transport.requestMapping(
-                                new MarshallerMappingItem(platformId, typeId, null),
-                                cache);
-
-                        clsName = fut.get().className();
-                    }
-                    else
-                        clsName = mappedName.className();
+            if (clientNode) {
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/MarshallerContextImpl.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientClusterImpl.java [hunk 7]
```diff
@@ -132,7 +132,7 @@ private boolean changeWalState(String cacheName, boolean enable) throws ClientEx
     private void checkClusterApiSupported(ProtocolContext protocolCtx)
         throws ClientFeatureNotSupportedByServerException {
         if (!protocolCtx.isFeatureSupported(ProtocolVersionFeature.CLUSTER_API) &&
-            !protocolCtx.isFeatureSupported(ProtocolBitmaskFeature.CLUSTER_API))
-            throw new ClientFeatureNotSupportedByServerException(ProtocolBitmaskFeature.CLUSTER_API);
+            !protocolCtx.isFeatureSupported(ProtocolBitmaskFeature.CLUSTER_ADDITIONAL_STATES))
+            throw new ClientFeatureNotSupportedByServerException(ProtocolBitmaskFeature.CLUSTER_ADDITIONAL_STATES);
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientClusterImpl.java]: This entire method looks incorrect to me: `CLUSTER_STATES` flag is only required for extra states, users should be able to use the API without this flag.

However, this is out of scope of the current PR. Too much stuff here as it is.

**Pull request merged**